### PR TITLE
Add annotation updating for migration for PVs and PVCs 

### PR DESCRIPTION
**Pull request by davidz627**
Add annotation annealing for migration for PVs and PVCs  on PV Controller startup. This allows external-provisioners to pick up and delete volumes when they have been rolled up from previous kubernetes versions.

Side note:
When externally provisioning the `volume.beta.kubernetes.io/storage-provisioner` annotation on the PVC was changed to the CSI Driver name during migration (since the introduction of CSI Migration feature). With this new addition of the `migrated-to` annotation it was possible to revert the `volume.beta.kubernetes.io/storage-provisioner` annotation to back to the in-tree plugin during migration for conceptual consistency. However, we determined that would be a breaking change to the Beta Migration feature and would cause additional headaches in terms of version compatability of external provisioners.

/assign @msau42 @misterikkit @ddebroy 
/cc @leakingtapan 

/kind bug
 
k8s/k8s part of solution for: #79043 

```release-note
Adds "volume.beta.kubernetes.io/migrated-to" annotation to PV's and PVC's when they are migrated to signal external provisioners to pick up those objects for Provisioning and Deleting. 
```

### CODE REVIEW

**Comment by jsafrane** [comment]
Can be the annotation removal added to `syncVolume` / `syncClaim` and adding the annotation to `provisionClaimOperationExternal()`? This way, the controller re-tries on error. On the other hand, it wastes a few CPU cycles to check the annotations and migrations in every sync.

**Comment by davidz627** [comment]
@jsafrane @misterikkit @msau42 this PR has been updated to match the revised design

**Comment by davidz627** [comment]
ping @msau42 @misterikkit @jsafrane for review :)

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 1]
```diff
@@ -309,6 +309,83 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
 }
 
+func (ctrl *PersistentVolumeController) updateClaimMigrationAnnotations(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// TODO: update[Claim|Volume]MigrationAnnotations can be optimized to not
+	// copy the claim/volume if no modifications are required. Though this
+	// requires some refactoring as well as an interesting change in the
+	// semantics of the function which may be undesirable. If no copy is made
+	// when no modifications are required this function could sometimes return a
+	// copy of the volume and sometimes return a ref to the original
+	copiedClaim := claim.DeepCopy()
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: for consistency with rest of file, can you call this `claimClone`?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 2]
```diff
@@ -244,6 +244,15 @@ type PersistentVolumeController struct {
 func (ctrl *PersistentVolumeController) syncClaim(claim *v1.PersistentVolumeClaim) error {
 	klog.V(4).Infof("synchronizing PersistentVolumeClaim[%s]: %s", claimToClaimKey(claim), getClaimStatusForLogging(claim))
 
+	// Set correct "migrated-to" annotations on PVC and update in API server if
+	// necessary
+	claim, err := ctrl.updateClaimMigrationAnnotations(claim)
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: Might need to check with @saad-ali but I think the space shuttle pattern has us assigning a new var `updatedClaim` and then explicitly writing `claim = updatedClaim` after the if block.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 3]
```diff
@@ -309,6 +309,83 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+	modified := annealMigrationAnnotations(ctrl.csiMigratedPluginManager, ctrl.translator, copiedClaim.Annotations, pvutil.AnnStorageProvisioner)
+	if modified {
+		newClaim, err := ctrl.kubeClient.CoreV1().PersistentVolumeClaims(copiedClaim.Namespace).Update(copiedClaim)
+		if err != nil {
+			return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations: %v", err)
+		}
+		_, err = ctrl.storeClaimUpdate(copiedClaim)
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: it looks like there is precedent for ignoring the return value here. Seems a little un-space-shuttle-like...

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 4]
```diff
@@ -309,6 +309,83 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+func (ctrl *PersistentVolumeController) updateClaimMigrationAnnotations(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// TODO: update[Claim|Volume]MigrationAnnotations can be optimized to not
+	// copy the claim/volume if no modifications are required. Though this
+	// requires some refactoring as well as an interesting change in the
+	// semantics of the function which may be undesirable. If no copy is made
+	// when no modifications are required this function could sometimes return a
+	// copy of the volume and sometimes return a ref to the original
+	copiedClaim := claim.DeepCopy()
+	modified := annealMigrationAnnotations(ctrl.csiMigratedPluginManager, ctrl.translator, copiedClaim.Annotations, pvutil.AnnStorageProvisioner)
+	if modified {
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: Even though it's space shuttle code, I think we're allowed to invert this condition and have an early return.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 5]
```diff
@@ -309,6 +309,83 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+	if !ok {
+		// Volume not dynamically provisioned. Ignore
+		return false
+	}
+
+	migratedToDriver := ann[pvutil.AnnMigratedTo]
+	if cmpm.IsMigrationEnabledForPlugin(provisioner) {
+		csiDriverName, err = translator.GetCSINameFromInTreeName(provisioner)
+		if err != nil {
+			klog.Errorf("Could not anneal volume migration annotations. Migration enabled for plugin %s but could not find corresponding driver name", provisioner)
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: please include `err` in the log.

**Review by misterikkit — COMMENTED** [review]
I think one thing missing from this PR is a comment (probably on `annealMigrationAnnotations` explaining the business logic for the new annotation. Is it as simple as,
- migration is on so we should put the csi driver name
- migration is off so we should delete the annotation
??

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: The `external-provisioner` is the entity that reads it off the `PVC`.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller_base.go]: brilliant, changed the name so we didn't shadow then shot myself in the foot :)

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 6]
```diff
@@ -309,6 +309,83 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+			return nil, fmt.Errorf("persistent Volume Controller can't anneal migration annotations: %v", err)
+		}
+		return newClaim, nil
+	}
+	return copiedClaim, nil
+}
+
+func (ctrl *PersistentVolumeController) updateVolumeMigrationAnnotations(volume *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	copiedVolume := volume.DeepCopy()
+	modified := annealMigrationAnnotations(ctrl.csiMigratedPluginManager, ctrl.translator, copiedVolume.Annotations, pvutil.AnnDynamicallyProvisioned)
```
[the remark this answers is not in the document]
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller_base.go]: the objects just store the same value in two different annotation keys
PV has `"pv.kubernetes.io/provisioned-by"`
and PVC has `"volume.beta.kubernetes.io/storage-provisioner"`
they essentially mean the same thing. I'm not sure why they're not the same key.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 7]
```diff
@@ -244,6 +244,15 @@ type PersistentVolumeController struct {
...
+	if err != nil {
+		// Nothing was saved; we will fall back into the same
+		// condition in the next call to this method
```
[the remark this answers is not in the document]
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: this is the same behavior as any other PV/PVC update failure - see call sites of things like `updateClaimStatus`, `bind`,  `updateVolumePhase`, etc. This update is of similar importance going forward - without an annotation there is no migration - if migration is the default there is no storage without the annotation.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 2 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: I don't really see this pattern elsewhere but it does seem to fit into the philosophy

File: pkg/controller/volume/persistentvolume/pv_controller_base.go (hunk 3 above)
Comment (davidz627) [reply pkg/controller/volume/persistentvolume/pv_controller_base.go]: I guess in these cases whether the cache was updated or not is not really relevant - doing an `if updated{foo()} else {foo()}` seems a little bit excessive

**Comment by davidz627** [comment]
@misterikkit addressed your comments, thanks!

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 2 above)
Comment (misterikkit) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Now I'm wondering where I saw it

File: pkg/controller/volume/persistentvolume/pv_controller_base.go (hunk 3 above)
Comment (misterikkit) [reply pkg/controller/volume/persistentvolume/pv_controller_base.go]: :man_shrugging:

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 8]
```diff
@@ -309,6 +309,90 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+	}
+	return newVol, nil
+
+}
+
+// updateMigrationAnnotations takes an Annotations map and checks for a
+// provisioner name using the provisionerKey. It will then add a
+// "volume.beta.kubernetes.io/migrated-to" annotation if migration with the CSI
+// driver name for that provisioner is "on" based on feature flags, it will also
+// remove the annotation is migration is "off" for that provisioner incd rollback
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: typo: incd

Please also explain the bool return value.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 9]
```diff
@@ -309,6 +309,90 @@ func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
...
+		}
+	} else {
+		if migratedToDriver != "" {
+			// Migration annotation exists but the driver isn't migrated currently
+			delete(ann, pvutil.AnnMigratedTo)
+			return true
+		}
+	}
+	return false
+
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: nit: extra blank line

File: pkg/controller/volume/persistentvolume/pv_controller_test.go [hunk 10]
```diff
@@ -461,3 +467,110 @@ func TestDelayBindingMode(t *testing.T) {
...
+		},
+	}
+
+	translator := csitrans.New()
+	cmpm := csimigration.NewPluginManager(translator)
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			for _, f := range tc.migratedDriverGates {
+				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, f, true)()
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_test.go]: hooray for global variables!

File: pkg/controller/volume/persistentvolume/pv_controller_test.go [hunk 11]
```diff
@@ -461,3 +467,110 @@ func TestDelayBindingMode(t *testing.T) {
...
+			expClaimAnnotations:  map[string]string{pvutil.AnnStorageProvisioner: gcePlugin, pvutil.AnnMigratedTo: gceDriver},
+			migratedDriverGates:  []featuregate.Feature{features.CSIMigrationGCE},
+		},
+		{
+			name:                 "migration off for GCE",
+			volumeAnnotations:    map[string]string{pvutil.AnnDynamicallyProvisioned: gcePlugin},
+			expVolumeAnnotations: map[string]string{pvutil.AnnDynamicallyProvisioned: gcePlugin},
+			claimAnnotations:     map[string]string{pvutil.AnnStorageProvisioner: gcePlugin},
+			expClaimAnnotations:  map[string]string{pvutil.AnnStorageProvisioner: gcePlugin},
+			migratedDriverGates:  []featuregate.Feature{},
```
Comment (misterikkit) [inline pkg/controller/volume/persistentvolume/pv_controller_test.go]: Will this test fail when GCE migration is on by default, and should we dump that problem on future us?

I vote yes.

**Review by misterikkit — COMMENTED** [review]
/lgtm

**Review by misterikkit — APPROVED** [review]
/lgtm

**Comment by davidz627** [comment]
These tests are crazy flaky - `Stage` stage fails repeatedly for `pull-kubernetes-e2e-gce` :'(
/retest

**Comment by davidz627** [comment]
Ok this is officially flaking too much to be considered a flake. It might be that this PR is making flakes worse or actually failing tests.
For my tracking purposes, these tests failed on the run that just ended:
```
Fail] [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir] [Testpattern: Pre-provisioned PV (default fs)] volumes [It] should store data 
[Fail] [sig-storage] EmptyDir volumes [It] should support (root,0666,tmpfs) [LinuxOnly] [NodeConformance] [Conformance] 
```
though looking at PR history logs show these tests failing the vast majority of the time in the past couple of days. Also the job history tool is super slow and buggy (showing a ton of `0001-01-01 00:00:00 +0000 UTC	0s` entries)

/cc @msau42 
in case you are aware of anything going on in the community around this

**Comment by msau42** [comment]
Message:OCI runtime start failed: container process is already dead

This is a known flake https://github.com/kubernetes/kubernetes/issues/86312

**Comment by davidz627** [comment]
/retest
Failed with Stage

**Comment by davidz627** [comment]
/retest
stage

**Comment by davidz627** [comment]
/retest
stage failure

**Comment by davidz627** [comment]
/retest
OCI RUNTIME

**Comment by davidz627** [comment]
/retest
STAGE!

**Comment by davidz627** [comment]
/hold
`[sig-storage] PersistentVolumes-local [Volume type: dir-bindmounted] One pod requesting one prebound PVC should be able to mount volume and write from pod1` failed in run https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/87098/pull-kubernetes-e2e-gce/1225092402105028609/

This might be a real problem (?) goign to search for known flake

**Comment by davidz627** [comment]
It's a flake. Opening an issue
/hold cancel

**Comment by davidz627** [comment]
/retest
STAGE
!!!

**Comment by davidz627** [comment]
hooray!

**Pull request merged**